WalletWasabi/Extensions/NBitcoinExtensions.cs

```diff
@@ -136,6 +137,55 @@ public static int GetAnonymitySet(this Transaction me, int outputIndex)
 
 		public static int GetAnonymitySet(this Transaction me, uint outputIndex) => GetAnonymitySet(me, (int)outputIndex);
 
+		public static int GetAnonymitySet(this Transaction me, int outputIndex, ICoinsView allWalletCoins)
+		{
+			var spentOwnCoins = allWalletCoins.OutPoints(me.Inputs.Select(x => x.PrevOut)).ToList();
+			var numberOfOwnInputs = spentOwnCoins.Count();
+			// If it's a normal tx that isn't self spent, nor a coinjoin, then anonymity should stripped if there was any and start from zero.
+			// Note: this is only a good idea from WWII, with WWI we calculate anonsets from the point the coin first hit the wallet.
+			// Note: a bit optimization to calculate this would be to actually use own output data, but that's a bit harder to get and this'll do it:
+			// If all our inputs are ours and there are more than 1 outputs then it's not a self-spent and it's not a coinjoin.
+			// This'll work, because we are only calculating anonset for our own coins and Wasabi doesn't generate tx that has more than one own outputs.
+			if (numberOfOwnInputs == me.Inputs.Count && me.Outputs.Count > 1)
+			{
+				return 1;
+			}
+
+			// Get the anonymity set of i-th output in the transaction.
+			var anonset = me.GetAnonymitySet(outputIndex);
+			// If we provided inputs to the transaction.
+			if (numberOfOwnInputs > 0)
+			{
+				// Take the input that we provided with the smallest anonset.
+				// Our smallest anonset input is the relevant here, because this way the common input ownership heuristic is considered.
+				var smallestInputAnon = spentOwnCoins.Min(x => x.AnonymitySet);
```

Comment:
> First, the spending transaction can contribute anon sets to each of the inputs because of their sibling inputs. Coins originating from transactions in which the wallet participated but where the outputs are not siblings of prevouts of the inputs can even have their anon set size estimated and added when accounting in inheritance computation if they are potentially related by non derived sub-transactions. More conservatively other coins that are relatable should be counted as contributing 1 to the input cluster anonymity set size.

Let me see if I get it:
Assuming sibling inputs to some batched transaction _b_ have CoinJoin histories, we want credit those add to the anon set of _b_'s outputs. I guess this is so change doesn't immediately drop to anonset = 1. Wallet-created change coins with at closest very distant relatives should add 1 to the anon set. I'm not sure exactly how "inheritance computation" is imagined here. The second sentence is not easy to read.

As it stands it's really hard for me to understand what's going on here.

> Setting a cap at the UI level (e.g. >9K) ...

 LOL

> Yet another hard upper bound, but one that is not not nearly conservative enough for privacy purposes can be estimated from the current size of the UTXO set.

which UTXO set